cli/auth.go

```diff
@@ -320,6 +320,180 @@ func (c *viamClient) createOrganizationAPIKey(orgID, keyName string) (*apppb.Cre
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	locationID := cCtx.String(dataFlagLocationID)
+	orgID := cCtx.String(dataFlagOrgID)
+	keyName := cCtx.String(apiKeyCreateFlagName)
+
+	if locationID == "" {
+		return errors.New("cannot create an api-key for a location without an ID")
+	}
+
+	if keyName == "" {
+		// Default name is in the form myusername@gmail.com-2009-11-10T23:00:00Z
+		// or key-uuid-2009-11-10T23:00:00Z if it was created by a key
+		keyName = fmt.Sprintf("%s-%s", c.conf.Auth, time.Now().Format(time.RFC3339))
+		infof(cCtx.App.Writer, "using default key name of %s", keyName)
+	}
+
+	if orgID == "" {
+		// try and create a key for the location based on the only org on the location
+		// if the user doesn't have perms for this org this won't work
+
+		loc, err := c.client.GetLocation(c.c.Context, &apppb.GetLocationRequest{
+			LocationId: locationID,
+		})
+		if err != nil {
+			return err
+		}
+
+		if loc.GetLocation() == nil {
+			return errors.Errorf("cannot create api-key for locationID: %s, empty location", locationID)
+		}
+
+		if len(loc.GetLocation().GetOrganizations()) != 1 {
+			return errors.Errorf("cannot create api-key for locationID: %s as there are mutiple orgs attached", locationID)
+		}
+
+		orgID = loc.GetLocation().GetOrganizations()[0].GetOrganizationId()
+		infof(cCtx.App.Writer, "creating key with default org on robot: %s", orgID)
+	}
+
+	req := &apppb.CreateKeyRequest{
+		Name: keyName,
+		Authorizations: []*apppb.Authorization{
+			{
+				AuthorizationType: "role",
+				AuthorizationId:   "location_owner",
+				ResourceType:      "location",
+				OrganizationId:    orgID,
+				IdentityType:      "api-key",
+				IdentityId:        "",
+			},
+		},
+	}
+
+	key, err := c.client.CreateKey(c.c.Context, req)
+	if err != nil {
+		return err
+	}
+
+	infof(cCtx.App.Writer, "Successfully created key: ")
+	printf(cCtx.App.Writer, "Key ID: %s \n", key.GetId())
+	printf(cCtx.App.Writer, "Key Value: %s \n", key.GetKey())
+	return nil
+}
+
+// RobotAPIKeyCreateAction corresponds to `robot api-key create`.
+func (c *viamClient) RobotAPIKeyCreateAction(cCtx *cli.Context) error {
+	c, err := newViamClient(cCtx)
+	if err != nil {
+		return err
+	}
+	err = c.robotAPIKeyCreateAction(cCtx)
+	return err
+}
+
+func (c *viamClient) robotAPIKeyCreateAction(cCtx *cli.Context) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	robotID := cCtx.String(dataFlagRobotID)
+	keyName := cCtx.String(apiKeyCreateFlagName)
+	orgID := cCtx.String(dataFlagOrgID)
+
+	if robotID == "" {
+		return errors.New("cannot create an api-key for a robot without an ID")
+	}
+
+	if keyName == "" {
+		// Default name is in the form myusername@gmail.com-2009-11-10T23:00:00Z
+		// or key-uuid-2009-11-10T23:00:00Z if it was created by a key
+		keyName = fmt.Sprintf("%s-%s", c.conf.Auth, time.Now().Format(time.RFC3339))
+		infof(cCtx.App.Writer, "using default key name of %s", keyName)
+	}
+
+	if orgID == "" {
```

Comment:
Like for LocationAPIKeyCreate action, I thought it would be cool to also do the "guess" the location if it wasn't provided by the user. This requires getting the location from the robot, and will fail if there are multiple orgs attached to the location. If the user who is trying to create a key for the robot does not perms for the location this will also fail. As we don't really support shared locations and keys right now I think this is quite a neat thing but I get that it is open to criticism so open to thoughts here.